ODIN_II/SRC/simulate_blif.cpp

```diff
@@ -1649,20 +1693,23 @@ static void compute_add_node(nnode_t* node, int cycle, int type) {
     oassert(node->num_input_port_sizes == 3);
     oassert(node->num_output_port_sizes == 2);
 
-    int i, num;
+    int i;
     int flag = 0;
 
     int* a = (int*)vtr::malloc(sizeof(int) * node->input_port_sizes[0]);
     int* b = (int*)vtr::malloc(sizeof(int) * node->input_port_sizes[1]);
     int* c = (int*)vtr::malloc(sizeof(int) * node->input_port_sizes[2]);
 
-    num = node->input_port_sizes[0] + node->input_port_sizes[1];
+    int num = node->input_port_sizes[0] + node->input_port_sizes[1];
     //if cin connect to unconn(PAD_NODE), a[0] connect to ground(GND_NODE) and b[0] connect to ground, flag = 0 the initial adder for addition
     //if cin connect to unconn(PAD_NODE), a[0] connect to ground(GND_NODE) and b[0] connect to vcc, flag = 1 the initial adder for subtraction
-    if (node->input_pins[num]->net->driver_pin->node->type == PAD_NODE) {
-        if (node->input_pins[0]->net->driver_pin->node->type == GND_NODE && node->input_pins[node->input_port_sizes[0]]->net->driver_pin->node->type == GND_NODE)
+#warning "Oops this be broke come fix"
```

Comment:
I believe you could resolve most of those to unknown automatically and throw a warning about wrongly formed circuit based on the blif standard (can't have a wire driven by more than one thing)... and given this is a blif simulator

Reply:
I thought the whole point of this was to support having a driver driven by more than one thing

Reply:
agreed... this is the blif simulator, I'm stating this should NOT error, just that the user is threading in invalid land and theres no specification for that so we are not gonna make up one

Reply:
also, printing unknown allows loosening the requirements on the vectors. if the expected input was unknown but the result is known, we only warn, we don't error unlike any other mismatch

Reply:
Im still quite confused. So just print a warning rather than assert and just assume 0?
Also, I haven't had a chance to test this thoroughly, but it seems to me that this would imply we cant use a variable that has been assigned to multiple times in an addition? - in which case do I need to rewrite this to support multiple drivers

Reply:
unless there is a better alternative?

Reply:
warn and assign -1 (unkown value) to the input_pin, this function evaluates the result based on input, if there is more than one, I can't pick, so just collapse into an unknown. technically everything should do that

Reply:
In theory you should be able to pick yes assuming one of the drivers is Z? I think there is probably some work here for managing that correctly. Currently I just fixed all the syntax errors to make it compile

Reply:
well yes, if the driver is disconnected it should, which is a problem, the simulator only has 0 1 and unknown, there is no disconnected value... we talked about this, but this will be a problem here.

Reply:
The only way to know wether its disconnected or not is to propagate the disconnected value. I think for know unknown is a temporary fix, but the simulator needs to change to quad value internally

Reply:
Ok I have added a warning and stated that behaviour is undefined and left a TODO. Are you happy with that resolution for the time being? Or do you want me to explicitly update all values to unknown

Reply:
I'd prefer update, its quick to do and it doesn't have a chance of being forgotten.

Reply:
I have a PR up that should give you quad valued logic hopefully soon.

Reply:
Ok hows that?